packages/electric-db-collection/src/electric.ts

```diff
@@ -239,67 +406,157 @@ export function electricCollectionOptions(
     })
   }
 
-  // Create wrapper handlers for direct persistence operations that handle txid awaiting
-  const wrappedOnInsert = config.onInsert
-    ? async (params: InsertMutationFnParams<any>) => {
-        // Runtime check (that doesn't follow type)
+  /**
+   * Wait for a custom match function to find a matching message
+   * @param matchFn Function that returns true when a message matches
+   * @param timeout Optional timeout in milliseconds (defaults to 3000ms)
+   * @returns Promise that resolves when a matching message is found
+   */
+  const awaitMatch: AwaitMatchFn<any> = async (
+    matchFn: MatchFunction<any>,
+    timeout: number = 3000
+  ): Promise<boolean> => {
+    debug(`awaitMatch called with custom function`)
 
-        const handlerResult =
-          ((await config.onInsert!(params)) as MaybeTxId) ?? {}
-        const txid = handlerResult.txid
+    return new Promise((resolve, reject) => {
+      const matchId = Math.random().toString(36)
+
+      const cleanupMatch = () => {
+        pendingMatches.setState((current) => {
+          const newMatches = new Map(current)
+          newMatches.delete(matchId)
+          return newMatches
+        })
+      }
+
+      const onTimeout = () => {
+        cleanupMatch()
+        reject(new TimeoutWaitingForMatchError())
+      }
 
-        if (!txid) {
-          throw new ElectricInsertHandlerMustReturnTxIdError()
+      const timeoutId = setTimeout(onTimeout, timeout)
+
+      // We need access to the stream messages to check against the match function
+      // This will be handled by the sync configuration
+      const checkMatch = (message: Message<any>) => {
+        if (matchFn(message)) {
+          debug(`awaitMatch found matching message, waiting for up-to-date`)
+          // Mark as matched but don't resolve yet - wait for up-to-date
+          pendingMatches.setState((current) => {
+            const newMatches = new Map(current)
+            const existing = newMatches.get(matchId)
+            if (existing) {
+              newMatches.set(matchId, { ...existing, matched: true })
+            }
+            return newMatches
+          })
+          return true
         }
+        return false
+      }
 
-        // Handle both single txid and array of txids
-        if (Array.isArray(txid)) {
-          await Promise.all(txid.map((id) => awaitTxId(id)))
-        } else {
-          await awaitTxId(txid)
+      // Check against current batch messages first to handle race conditions
+      for (const message of currentBatchMessages.state) {
+        if (checkMatch(message)) {
```

Comment:
Why is `checkMatch` looking up the match and setting `match.matched` to `true` if we are also doing that here on L465? Isn't L465 going to override whatever we set on L449?